File: CurrencyAssistantServer/CurrencyAssistantServer/Push/Network/NetworkManager.m
//
//  NetworkManager.m
//  SmartPush
//
//  Created by shao on 2021/8/24.
//  Copyright © 2021 www.skyfox.org. All rights reserved.
//

#import "NetworkManager.h"

@implementation NetworkManager

#pragma mark - public methods

+ (NetworkManager *)sharedManager {
    static NetworkManager *sharedManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedManager = [[NetworkManager alloc] init];
    });
    return sharedManager;
}

- (void)setIdentity:(SecIdentityRef)identity {
    if (_identity == identity) {
        return;
    }
    if (_identity != NULL) {
        CFRelease(_identity);
    }
    if (identity != NULL) {
        _identity = (SecIdentityRef)CFRetain(identity);
        // Create a new session
        NSURLSessionConfiguration *conf = [NSURLSessionConfiguration ephemeralSessionConfiguration];
        self.session = [NSURLSession sessionWithConfiguration:conf
                                                     delegate:self
                                                delegateQueue:[NSOperationQueue mainQueue]];
    } else {
        _identity = NULL;
    }
}

- (void)postWithPayload:(NSString *)payload
                toToken:(NSString *)token
              withTopic:(nullable NSString *)topic
               priority:(NSUInteger)priority
             collapseID:(NSString *)collapseID
            payloadType:(NSString *)payloadType
              inSandbox:(BOOL)sandbox
             exeSuccess:(void(^)(id responseObject))exeSuccess
              exeFailed:(void(^)(NSString *error))exeFailed {
    
    NSString *url = [NSString stringWithFormat:@"https://api%@.push.apple.com/3/device/%@",
                     sandbox ? @".sandbox" : @"",
                     token];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:url]];
    request.HTTPMethod = @"POST";
    request.HTTPBody = [payload dataUsingEncoding:NSUTF8StringEncoding];
    if (topic) {
        [request addValue:topic forHTTPHeaderField:@"apns-topic"];
    }
    if (collapseID.length > 0) {
        [request addValue:collapseID forHTTPHeaderField:@"apns-collapse-id"];
    }
    NSString *priorityStr = [NSString stringWithFormat:@"%lu", (unsigned long)priority];
    [request addValue:priorityStr forHTTPHeaderField:@"apns-priority"];
    [request addValue:payloadType forHTTPHeaderField:@"apns-push-type"];
    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request
                                                 completionHandler:^(NSData * _Nullable data,
                                                                     NSURLResponse * _Nullable response,
                                                                     NSError * _Nullable error) {
        NSHTTPURLResponse *r = (NSHTTPURLResponse *)response;
        if (r == nil || error) {
            if (exeFailed) {
                exeFailed(error.debugDescription);
            }
            return;
        }
        NSError *perror;
        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data
                                                             options:0
                                                               error:&perror];
        if (r.statusCode == 200 && !error) {
            if (exeSuccess) {
                exeSuccess(dict);
            }
            return;
        }
        NSString *reason = dict[@"reason"];
        NSLog(@"reason:%@", reason);
        exeFailed(reason);
    }];
    [task resume];
}

#pragma mark - NSURLSessionDelegate

- (void)  URLSession:(NSURLSession *)session
                task:(nonnull NSURLSessionTask *)task
 didReceiveChallenge:(nonnull NSURLAuthenticationChallenge *)challenge
   completionHandler:(nonnull void (^)(NSURLSessionAuthChallengeDisposition,
                                       NSURLCredential * _Nullable))completionHandler {
    SecCertificateRef certificate;
    SecIdentityCopyCertificate(self.identity, &certificate);
    NSURLCredential *cred = [[NSURLCredential alloc] initWithIdentity:self.identity
                                                         certificates:@[(__bridge_transfer id)certificate]
                                                          persistence:NSURLCredentialPersistenceForSession];
    completionHandler(NSURLSessionAuthChallengeUseCredential, cred);
}

@end

File: CurrencyAssistantServer/CurrencyAssistantServer/Push/Network/NetworkManager.swift
//
//  NetworkManager.swift
//  CurrencyAssistantServer
//
//  Created by neilson on 2022-03-30.
//

import Foundation

class NetworkManager : NSObject {
    
    public static let shared = NetworkManager()
    public var identity: SecIdentity? {
        didSet {
            let conf = URLSessionConfiguration.ephemeral
            session = URLSession.init(configuration: conf,
                                      delegate: self,
                                      delegateQueue: .main)
        }
    }
    private var session: URLSession?

    /*
     * https://developer.apple.com/documentation/usernotifications
     * /setting_up_a_remote_notification_server
     * /sending_notification_requests_to_apns?language=objc
     */
    public func postWithPayload(_ payload: String,
                                token: String,
                                topic: String?,
                                priority: UInt,
                                collapseID: String?,
                                payloadType: String,
                                sandBox: Bool,
                                exeSuccess: @escaping ([String: Any]?) -> Void,
                                exeFailed: @escaping (String) -> Void) {
        let subUrlString = sandBox ? ".sandbox" : ""
        guard let url = URL(string: "https://api\(subUrlString).push.apple.com/3/device/\(token)") else {
            return
        }
        var request = URLRequest.init(url: url)
        request.httpMethod = "POST"
        request.httpBody = payload.data(using: .utf8)
        if let t = topic, !t.isEmpty {
            request.addValue(t, forHTTPHeaderField: "apns-topic")
        }
        if let c = collapseID, !c.isEmpty {
            request.addValue(c, forHTTPHeaderField: "apns-collapse-id")
        }
        request.addValue(String(priority), forHTTPHeaderField: "apns-priority")
        request.addValue(payloadType, forHTTPHeaderField: "apns-push-type")
        let task = session?.dataTask(with: request,
                                     completionHandler: { (data, response, error) in
            guard error == nil, let r = response as? HTTPURLResponse else {
                exeFailed(error.debugDescription)
                return
            }
            
            var result: [String: Any]?
            if let data = data, !data.isEmpty {
                do {
                    result = try JSONSerialization.jsonObject(with: data, options: []) as? Dictionary<String, Any>
                } catch {
                    print("reason:\(error.localizedDescription)")
                    exeFailed(error.localizedDescription)
                    return
                }
            }

            if r.statusCode == 200 {
                exeSuccess(result)
                return
            }
            let reason = result?["reason"] as? String ?? "undefined error"
            print("reason:\(reason)")
            exeFailed(reason)
        })
        task?.resume()
    }
}

extension NetworkManager : URLSessionDelegate {
    func urlSession(_ session: URLSession,
                    didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard let id = identity else {
            return
        }
        var cert: SecCertificate?
        SecIdentityCopyCertificate(id, &cert)
        guard let c = cert else {
            return
        }
        let cred = URLCredential.init(identity: id,
                                      certificates: [c],
                                      persistence: .forSession)
        completionHandler(.useCredential, cred)
    }
}
